cassandra/cluster.py

```diff
@@ -3118,36 +3152,27 @@ def prepare(self, query, custom_payload=None, keyspace=None):
         **Important**: PreparedStatements should be prepared only once.
         Preparing the same query more than once will likely affect performance.
 
+        When :meth:`~.Cluster.prepare_on_all_hosts` is enabled, method
+        attempts to prepare given query on all hosts and waits for each node to respond.
+        Preparing CQL query on other nodes may fail, but error is not propagated
+        to the caller.
+
         `custom_payload` is a key value map to be passed along with the prepare
         message. See :ref:`custom_payload`.
         """
-        message = PrepareMessage(query=query, keyspace=keyspace)
-        future = ResponseFuture(self, message, query=None, timeout=self.default_timeout)
-        try:
-            future.send_request()
-            response = future.result().one()
-        except Exception:
-            log.exception("Error preparing query:")
-            raise
-
-        prepared_keyspace = keyspace if keyspace else None
-        prepared_statement = PreparedStatement.from_message(
-            response.query_id, response.bind_metadata, response.pk_indexes, self.cluster.metadata, query, prepared_keyspace,
-            self._protocol_version, response.column_metadata, response.result_metadata_id, self.cluster.column_encryption_policy)
-        prepared_statement.custom_payload = future.custom_payload
-
-        self.cluster.add_prepared(response.query_id, prepared_statement)
-
+        future = self.prepare_async(query, custom_payload, keyspace, prepare_on_all_hosts=False)
+        response = future.result()
         if self.cluster.prepare_on_all_hosts:
+            # prepare on all hosts in a synchronous way, not asynchronously
+            # as internally in prepare_async() (PrepareFuture)
             host = future._current_host
             try:
-                self.prepare_on_all_hosts(prepared_statement.query_string, host, prepared_keyspace)
+                self.prepare_on_all_nodes(response.query_string, host, response.keyspace)
             except Exception:
                 log.exception("Error preparing query on all hosts:")
+        return response
```

Comment:
This seems exactly right: prepare() should be implemented as prepare_async().get() once we have a good working prepare_async().  But why do we need to do the prepare_on_all_nodes() operation here?  We should've already done that when future.result() completed (since it's done in PrepareFuture._set_final_result()).